### DISCUSSION TITLE
Proposal: Allow `Func<T, T, bool>` where `IEqualityComparer<T>` is expected

### DISCUSSION THREAD

**Discussion by dersia** [dotnet/csharplang#8583] [Language Ideas] [answered]
I saw a Api design in the runtime repo, where there are bunch of methods on `MemoryExtensions` are receiving an `IEqualityComparer<T>` overload so that `T` can be unconstrained on those methods.

this made me think, if it would be a good idea to allow `Func<T, T, bool>` to be used as parameters where an `IEqualityComparer<T>` is expected. This would allow a scenario where for
```c#
public static bool Contains<T>(this ReadOnlySpan<T> span, T value, IEqualityComparer<T>? comparer = null);
```
the following would be allowed:
```c#
var c = [1, 2, 3, 4, 5];

c.Contains(5, (a, b) => a == b);
```

(I know what this is a very simple example and can be solved differently, but I hope the idea is clear.)

this would be implemented by emitting a anonymous class that implements the interface and uses the `Func<T, T, bool>` in the equals method.
this would mean, that I don't have to create comparer classes just for those use cases, where I am not reusing the comparer for anything else.

I am not sure, if I am the only one with this issue, but I guess, this would simplify a lot of codebases.

**Comment by HaloFour** [comment]
See: #2517

**Comment by ufcpp** [comment]
And https://github.com/dotnet/csharplang/discussions/728

**Comment by TahirAhmadov** [comment, marked as answer]
That interface has `Equals` and `GetHashCode`, does the language have to run around and look for matching signatures? What if there are 2 methods with identical signatures? I think this is a bad idea. 
Declaring a class like `DelegateEqualityComparer<T>` is trivial (and it's probably already available in some general purpose helper Nuget).

**Reply by Trarizon** [reply]
Actually it exists in BCL. [EqualityComparer<T>.Create](https://github.com/dotnet/runtime/blob/756f7ebbbdcc0b8cb8b383ef3b6e6b50430f1922/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs#L28)

**Reply by dersia** [reply]
I wasn't aware of IEqualityComparer.Create. thanks for that.

@TahirAhmadov for some reason it slipped my mind to think of a general compare class, that could be used that way. thanks for pointing at the obvious solution. 🤦
